src/new_recur.jl

```diff
@@ -0,0 +1,83 @@
+
+
+"""
+  NewRecur
+New Recur. An experimental recur interface for removing statefullness in recurrent architectures for flux.
+"""
+struct NewRecur{RET_SEQUENCE, T}
+  cell::T
+  # state::S
+  function NewRecur(cell; return_sequence::Bool=false)
+    new{return_sequence, typeof(cell)}(cell)
+  end
+  function NewRecur{true}(cell)
+    new{true, typeof(cell)}(cell)
+  end
+  function NewRecur{false}(cell)
+    new{false, typeof(cell)}(cell)
+  end
+end
+
+# This is the same way we do 3-tensers from Flux.Recur
+function (m::NewRecur{false})(x::AbstractArray{T, N}, carry) where {T, N}
+  @assert N >= 3
+  # h = [m(x_t) for x_t in eachlastdim(x)]
+
+  cell = l.cell
+  x_init, x_rest = Iterators.peel(xs)
+  (carry, y) = cell(carry, x_init)
+  for x in x_rest
+    (carry, y) = cell(carry, x)
+  end
+  # carry, y
+  y
+
+end
+
+function (l::NewRecur{false})(x::AbstractArray{T, 3}, carry=l.cell.state0) where T
+  m(Flux.eachlastdim(x), carry)
+end
+
+function (l::NewRecur{false})(xs::Union{AbstractVector{<:AbstractArray}, Base.Generator},
+                              carry=l.cell.state0)
+  rnn = l.cell
+  # carry = layer.stamte
+  x_init, x_rest = Iterators.peel(xs)
+  (carry, y) = rnn(carry, x_init)
+  for x in x_rest
+    (carry, y) = rnn(carry, x)
+  end
+  y
+end
+
+# From Lux.jl: https://github.com/LuxDL/Lux.jl/pull/287/
+function (l::NewRecur{true})(xs::Union{AbstractVector{<:AbstractArray}, Base.Generator},
+                             carry=l.cell.state0)
+  rnn = l.cell
+  _xs = if xs isa Base.Generator
+    collect(xs)  # TODO: Fix. I can't figure out how to get around this for generators.
+  else
+    xs
+  end
```

Comment:
It's fine for us to not accept `Generator`s and other arbitrary iterables. I'd be fine with even rejecting 1- and 2D arrays for a first prototype, or accepting `Vector{<:AbstractArray}` and `AbstractArray{..., N}` but erroring on `AbstractMatrix`.